[Ingest] Add `revision` to agent configs & data sources

jen-huang opened the pull request:
Partially resolves #56750

## Summary

This PR adds a `revision` property to agent config saved objects. The value starts at `1` when a new config is created and increments whenever it is updated, including when data sources are created/updated/deleted from the config. A data source deletion route & service was also added to be able to test everything properly.

I also added the `revision` property to data source saved objects and it increments in the same manner. This isn't bubbled up to the UI anywhere yet, but it may be useful for concurrency control in the future.

This property is implemented as a custom saved object attribute, instead of leveraging the internal `version` attribute, because as of ES 7.0, there's no monotonically increasing version number ([more about this here](https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-7.0.html#_internal_versioning_is_no_longer_supported_for_optimistic_concurrency_control)).

**Note:** you may need to clear your ES data to test this PR as the saved objects mapping was changed.

## Screenshot
Agent config list UI:
![image](https://user-images.githubusercontent.com/1965714/76366749-1c013f80-62e8-11ea-8cca-03e0c939640f.png)

elasticmachine commented:
Pinging @elastic/ingest-management (Feature:Fleet)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32284/)
* Commit: b68e0ab90d0e09f6fd4f7f7a1b33b666fc8471a8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32284/","number":"32284","commit":"b68e0ab90d0e09f6fd4f7f7a1b33b666fc8471a8"}]}
PIPELINE-->

nchaulet commented:
it would be nice to add the running config revision to the agent too, I think we could get it when an agent ack after a config change  (but it could be done in a different PR)

nchaulet commented on x-pack/plugins/ingest_manager/server/services/datasource.ts:
```diff
@@ -4,23 +4,34 @@
+      {
+        ...datasource,
+        revision: 1,
+      } as Datasource,
```
Kind of incorrect, this is not a Datasource there is no id, for agent I have a different type `AgentSOAttributes` for saved object maybe we can use the same pattern here?

jen-huang replied:
updated typings here

nchaulet commented on x-pack/plugins/ingest_manager/server/services/datasource.ts:
```diff
@@ -98,14 +109,61 @@ class DatasourceService {
+    });
+
+    // Bump revision of associated agent config
+    await agentConfigService.update(soClient, datasource.config_id, {}, { user: options?.user });
```
maybe we can have a method with a more explicit name and avoid the comment  `agentConfigService.bumpRevision` ? or `agentConfigService.datasourceUpdate`?

jen-huang replied:
added `bumpRevision`

nchaulet commented on x-pack/plugins/ingest_manager/server/routes/datasource/index.ts:
```diff
@@ -58,4 +60,14 @@ export const registerRoutes = (router: IRouter) => {
+    {
+      path: DATASOURCE_API_ROUTES.DELETE_PATTERN,
+      validate: DeleteDatasourcesRequestSchema,
+      options: { tags: [`access:${PLUGIN_ID}`] },
```
It's not implemented anywhere for now, but we should make a difference between read and write API routes here it should be only accessible for user with write access: `access:ingestManager_write`

jen-huang replied:
I think this would be good to fix across all routes, deferring for now

jen-huang commented:
> it would be nice to add the running config revision to the agent too, I think we could get it when an agent ack after a config change (but it could be done in a different PR)

@nchaulet Agreed, I started looking at the agent model and various services to add this, but as I'm not the most familiar with all pieces, I think it would be faster if you took on this work? I updated #56750 to include tasks for agent side of things and updated this PR description to be partial resolution of that issue.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32900/)
* Commit: 5e3dfaa3e36712807d760cebcbff8c8c83a6167a

### History
* :green_heart: [Build #32284](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32284/) succeeded b68e0ab90d0e09f6fd4f7f7a1b33b666fc8471a8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32284/","number":"32284","commit":"b68e0ab90d0e09f6fd4f7f7a1b33b666fc8471a8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/32900/","number":"32900","commit":"5e3dfaa3e36712807d760cebcbff8c8c83a6167a"}]}
PIPELINE-->
